src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileParser.cs

```diff
@@ -20,28 +20,84 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
+			{
+				IncludeXamlNamespaces = includeXamlNamespaces;
+				ExcludeXamlNamespaces = excludeXamlNamespaces;
+				File = file;
+				ContentsHash = contentsHash;
+			}
+
 			public string[] IncludeXamlNamespaces { get; }
 			public string[] ExcludeXamlNamespaces { get; }
-			public DateTime LastTimeUsed { get; }
+			public string File { get; }
+			public int ContentsHash { get; }
+
+			public override bool Equals(object? obj)
+				=> obj is CachedFileKey key && Equals(key);
+
+			public bool Equals(CachedFileKey other)
+			{
+				return EqualityComparer<string[]>.Default.Equals(IncludeXamlNamespaces, other.IncludeXamlNamespaces) &&
+					EqualityComparer<string[]>.Default.Equals(ExcludeXamlNamespaces, other.ExcludeXamlNamespaces) &&
+					File == other.File &&
+					ContentsHash == other.ContentsHash;
+			}
+
+			public override int GetHashCode()
+			{
+				var hashCode = -1678229606;
+				hashCode = hashCode * -1521134295 + EqualityComparer<string[]>.Default.GetHashCode(IncludeXamlNamespaces);
+				hashCode = hashCode * -1521134295 + EqualityComparer<string[]>.Default.GetHashCode(ExcludeXamlNamespaces);
+				hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(File);
+				hashCode = hashCode * -1521134295 + ContentsHash.GetHashCode();
+				return hashCode;
+			}
+
+			public static bool operator ==(CachedFileKey left, CachedFileKey right) => left.Equals(right);
+			public static bool operator !=(CachedFileKey left, CachedFileKey right) => !(left == right);
+		}
+
+		private struct CachedFile : IEquatable<CachedFile>
+		{
+			public DateTimeOffset LastTimeUsed { get; }
 			public XamlFileDefinition XamlFileDefinition { get; }
 
-			public CachedFile(string[] includeXamlNamespaces, string[] excludeXamlNamespaces, DateTime lastTimeUsed, XamlFileDefinition xamlFileDefinition)
+			public CachedFile(DateTimeOffset lastTimeUsed, XamlFileDefinition xamlFileDefinition)
 			{
-				IncludeXamlNamespaces = includeXamlNamespaces;
-				ExcludeXamlNamespaces = excludeXamlNamespaces;
 				LastTimeUsed = lastTimeUsed;
 				XamlFileDefinition = xamlFileDefinition;
 			}
 
 			public CachedFile WithUpdatedLastTimeUsed()
 			{
-				return new CachedFile(IncludeXamlNamespaces, ExcludeXamlNamespaces, DateTime.Now, XamlFileDefinition);
+				return new CachedFile(DateTimeOffset.Now, XamlFileDefinition);
 			}
+
+			public override bool Equals(object? obj)
+				=> obj is CachedFile file && Equals(file);
+
+			public bool Equals(CachedFile other)
+			{
+				return LastTimeUsed.Equals(other.LastTimeUsed) &&
+					EqualityComparer<XamlFileDefinition>.Default.Equals(XamlFileDefinition, other.XamlFileDefinition);
+			}
+
+			public override int GetHashCode()
+			{
+				var hashCode = 368878473;
+				hashCode = hashCode * -1521134295 + LastTimeUsed.GetHashCode();
+				hashCode = hashCode * -1521134295 + EqualityComparer<XamlFileDefinition>.Default.GetHashCode(XamlFileDefinition);
+				return hashCode;
```

Comment:
Which implementation is this based on? (the numbers specifically)

Reply:
Roslyn feature in Visual Studio to generate Equals and GetHashCode. Just <kbd>Ctrl</kbd>+<kbd>.</kbd> on struct name that doesn't have any equality contracts (interface, operators, object Equals override, and object GetHashCode override), and VS will generate this.

Reply:
fwiw, it seems that `-1521134295` is hard-coded in Roslyn, while the initial `hashCode` value (`368878473`) is not hard-coded.

Feature tests are in https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/GenerateFromMembers/GenerateEqualsAndGetHashCodeFromMembers/GenerateEqualsAndGetHashCodeFromMembersTests.cs